### PR TITLE
Add support for Direct Workload Identity auth

### PR DESCRIPTION
**Pull request by sethvargo**
This adds a new authentication mode, Direct Workload Identity Federation. This new mode permits authenticating to Google Cloud directly using the GitHub Actions OIDC token instead of proxying through a Google Cloud Service Account.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -287,256 +288,163 @@ regardless of the authentication mechanism.
 -   `access_token`: The Google Cloud access token for calling other Google Cloud
     APIs. This is only available when "token_format" is "access_token".
 
--   `access_token_expiration`: The RFC3339 UTC "Zulu" format timestamp for the
```
Comment (verbanicm) [inline README.md:290]: Should `auth_token` be documented in the outputs?

**Review by verbanicm — DISMISSED** [review]
One comment about missing output documentation for `auth_token`

**Review by verbanicm — APPROVED** [review]

File: docs/EXAMPLES.md [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+jobs:
+  job_id:
+    permissions:
+      contents: 'read'
+      id-token: 'write'
+
+    - id: 'auth'
+      uses: 'google-github-actions/auth@v2'
+      with:
+        project_id: 'my-project'
```
Comment (squee1945) [inline docs/EXAMPLES.md:25]: `project_id` is required when `service_account` is not specified?

File: docs/EXAMPLES.md [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+      contents: 'read'
+      id-token: 'write'
+
+    steps:
+    - uses: 'actions/checkout@v4'
+
+    - id: 'auth'
+      uses: 'google-github-actions/auth@v2'
+      with:
+        workload_identity_provider: 'projects/123456789/locations/global/workloadIdentityPools/my-pool/providers/my-provider'
```
[the remark this answers is not in the document]
Comment (sethvargo) [reply docs/EXAMPLES.md:104]: It's required when we can't extract it from any of the inputs.

**Comment by dghubble** [comment]
Is there further reading here? The direct vs service princiapl distinction isn't mentioned anywhere I've seen

**Comment by squee1945** [comment]
It's a new feature and it seems to be mostly undocumented at the moment. It is implied by the documentation here: https://cloud.google.com/iam/docs/principal-identifiers#v1\

Effectively, it is now possible to authorize a workload/workforce identity pool principal directly, without the need to impersonate a service account. It is broadly supported across GCP products, but not yet completely, which may explain the lack of documentation at the moment.

**Comment by skikkh** [comment]
> However, not all Google Cloud resources support principalSet identities.

I am currently seeking clarification regarding the specific Google Cloud resources that do not support principalSet identities. Unfortunately, I have not been able to find detailed information on this topic in the official Google Cloud documentation.
The README.md states it as 'Preferred', and I am interested in using it. However, I am having trouble because the range of resources that can be utilized remains unclear.

**Comment by glasser** [comment]
After this PR, this bit in the README is misleading/incomplete:

> project_id: (Optional) Custom project ID to use for authentication and exporting into other steps. If unspecified, the project ID will be extracted from the Workload Identity Provider or the Service Account Key JSON.

This makes it seem like if you specify `workload_identity_provider` by itself, it will be able to extract the project ID, but that seems to not be the case — it only knows how to get it from the service account email address (which is now option), not the `workload_identity_provider` itself. I think this is just a doc issue (you can't get the project ID from the `workload_identity_provider` because project ID is a name and `workload_identity_provider` seems to require the project number).

**Comment by sethvargo** [comment]
> After this PR, this bit in the README is misleading/incomplete...

https://github.com/google-github-actions/auth/pull/367

**Pull request merged**